File: artifacts/api-server/src/lib/topic-security.ts
const MAX_TOPIC_LENGTH = 200;
const ALLOWED_TOPIC_CHARACTER = /^[\p{L}\p{N}\s&+.,:;/'()\-#]$/u;
const CONTROL_OR_FORMAT_CHARACTER = /[\u0000-\u001f\u007f-\u009f\u200b-\u200f\u202a-\u202e\u2060-\u2064\u2066-\u2069\ufeff]/u;

const PROMPT_INJECTION_PATTERNS = [
  /\bignore(?:\s+(?:all|any|the))?\s+(?:(?:previous|prior|above|earlier)\s+)?instructions?\b/i,
  /\b(?:disregard|override|forget|bypass)\b.{0,40}\b(?:instructions?|rules?|prompt|system\s+message)\b/i,
  /\b(?:system|developer)\s+(?:prompt|message|instructions?)\b/i,
  /\b(?:reveal|show|print|repeat|output)\s+(?:the\s+)?(?:system|developer)\s+(?:prompt|message|instructions?)\b/i,
  /\bdo\s+not\s+follow\b/i,
] as const;

export class TopicValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TopicValidationError";
  }
}

export function sanitizeResearchTopic(value: unknown): string {
  if (typeof value !== "string") {
    throw new TopicValidationError("Research topic is required.");
  }

  const normalized = value.normalize("NFKC").trim().replace(/\s+/gu, " ");
  if (!normalized) {
    throw new TopicValidationError("Research topic is required.");
  }
  if (normalized.length > MAX_TOPIC_LENGTH) {
    throw new TopicValidationError(
      `Research topic must be ${MAX_TOPIC_LENGTH} characters or fewer.`,
    );
  }
  if (CONTROL_OR_FORMAT_CHARACTER.test(normalized)) {
    throw new TopicValidationError(
      "Research topic contains unsupported invisible characters.",
    );
  }
  if (PROMPT_INJECTION_PATTERNS.some((pattern) => pattern.test(normalized))) {
    throw new TopicValidationError(
      "Enter a research subject only; instruction-like text is not allowed.",
    );
  }
  if (
    [...normalized].some(
      (character) => !ALLOWED_TOPIC_CHARACTER.test(character),
    )
  ) {
    throw new TopicValidationError(
      "Use words, numbers, spaces, and common research punctuation only.",
    );
  }

  return normalized;
}

File: artifacts/abstractly/backend/abstractly/topic_security.py
"""Validation for user-provided research topics."""

from __future__ import annotations

import re
import unicodedata


MAX_TOPIC_LENGTH = 200
_ALLOWED_TOPIC_CHARACTERS = frozenset("&+.,:;/'()-#")
_CONTROL_OR_FORMAT_CATEGORIES = frozenset({"Cc", "Cf", "Cs", "Co", "Cn"})
_PROMPT_INJECTION_PATTERNS = (
    re.compile(
        r"\bignore(?:\s+(?:all|any|the))?\s+"
        r"(?:(?:previous|prior|above|earlier)\s+)?instructions?\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:disregard|override|forget|bypass)\b.{0,40}"
        r"\b(?:instructions?|rules?|prompt|system\s+message)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:system|developer)\s+(?:prompt|message|instructions?)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:reveal|show|print|repeat|output)\s+(?:the\s+)?"
        r"(?:system|developer)\s+(?:prompt|message|instructions?)\b",
        re.IGNORECASE,
    ),
    re.compile(r"\bdo\s+not\s+follow\b", re.IGNORECASE),
)


class TopicValidationError(ValueError):
    """Raised when a topic is empty, unsafe, or unsupported."""


def sanitize_research_topic(value: object) -> str:
    """Normalize a topic and reject prompt-like or unusual input."""
    if not isinstance(value, str):
        raise TopicValidationError("Research topic is required.")

    normalized = " ".join(
        unicodedata.normalize("NFKC", value).strip().split()
    )
    if not normalized:
        raise TopicValidationError("Research topic is required.")
    if len(normalized) > MAX_TOPIC_LENGTH:
        raise TopicValidationError(
            f"Research topic must be {MAX_TOPIC_LENGTH} characters or fewer."
        )
    if any(
        unicodedata.category(character) in _CONTROL_OR_FORMAT_CATEGORIES
        for character in normalized
    ):
        raise TopicValidationError(
            "Research topic contains unsupported invisible characters."
        )
    if any(pattern.search(normalized) for pattern in _PROMPT_INJECTION_PATTERNS):
        raise TopicValidationError(
            "Enter a research subject only; instruction-like text is not allowed."
        )
    if any(
        not (character.isalnum() or character.isspace() or character in _ALLOWED_TOPIC_CHARACTERS)
        for character in normalized
    ):
        raise TopicValidationError(
            "Use words, numbers, spaces, and common research punctuation only."
        )

    return normalized